src/dex/bebop/bebop.ts

```diff
@@ -0,0 +1,744 @@
+        }, ${side}:`,
+        e,
+      );
+      return null;
+    }
+  }
+
+  // Returns estimated gas cost of calldata for this DEX in multiSwap
+  getCalldataGasCost(poolPrices: PoolPrices<BebopData>): number | number[] {
+    // This relies heavily on exact quote. Can we note use Bebop Data to find this? either via pools or the data itself?
+    // This assumes that a single maker was used to fill this trade
+    // "order":{
+    return (
+      CALLDATA_GAS_COST.DEX_OVERHEAD +
+      // "expiry":"1725541751"
+      CALLDATA_GAS_COST.TIMESTAMP +
+      // "taker_address":"0x9008d19f58aabd9ed0d60971565aa8510560ab41"
+      CALLDATA_GAS_COST.ADDRESS +
+      // "maker_address":"0x807cf9a772d5a3f9cefbc1192e939d62f0d9bd38"
+      CALLDATA_GAS_COST.ADDRESS +
+      // "maker_nonce":"1725541661402362944"
+      CALLDATA_GAS_COST.UUID +
+      // "taker_token":"0x3429d03c6f7521aec737a0bbf2e5ddcef2c3ae31"
+      CALLDATA_GAS_COST.ADDRESS +
+      // "maker_token":"0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2"
+      CALLDATA_GAS_COST.ADDRESS +
+      // "taker_amount":"3789033775358849793876"
+      CALLDATA_GAS_COST.AMOUNT +
+      // "maker_amount":"199726023273987970"
+      CALLDATA_GAS_COST.AMOUNT +
+      // "receiver":"0x9008d19f58aabd9ed0d60971565aa8510560ab41"
+      CALLDATA_GAS_COST.ADDRESS +
+      // "packed_commands":"0"
+      CALLDATA_GAS_COST.wordNonZeroBytes(4) +
+      // "flags":"103712057072216206793253728668500811363805133225332165859664901734817333248000"
+      CALLDATA_GAS_COST.FULL_WORD +
+      // }
+      // "makerSignature":{
+      // "signatureBytes":"0x34e1fbda1d48296949f0a003d4194646ed750a4ccd408b52979f60c2e7332207190a91600a40396b25fd645422dde85bc0ff53f26b66202a985ac8b9da87b47e1c"
+      CALLDATA_GAS_COST.FULL_WORD +
+      // "flags":"1"
+      CALLDATA_GAS_COST.LENGTH_SMALL +
+      // }
+      // "filledTakerAmount":"0"
+      CALLDATA_GAS_COST.AMOUNT
+    );
+  }
+
+  // Encode params required by the exchange adapter
+  // V5: Used for multiSwap, buy & megaSwap
+  // V6: Not used, can be left blank
+  // Hint: abiCoder.encodeParameter() could be useful
+  getAdapterParam(
+    srcToken: string,
+    destToken: string,
+    srcAmount: string,
+    destAmount: string,
+    data: BebopData,
+    side: SwapSide,
+  ): AdapterExchangeParam {
+    const { tx } = data;
+
+    if (!tx) {
+      throw new Error('No tx in data');
+    }
+
+    // Encode here the payload for adapter
+    const payload = tx.data;
+
+    return {
+      targetExchange: this.settlementAddress,
+      payload,
+      networkFee: '0',
+    };
+  }
+
+  // This is called once before getTopPoolsForToken is
+  // called for multiple tokens. This can be helpful to
+  // update common state required for calculating
+  // getTopPoolsForToken. It is optional for a DEX
+  // to implement this
+  async updatePoolState(): Promise<void> {
+    const tokens = await this.getCachedTokens();
+
+    if (tokens) {
+      this.tokensMap = tokens;
+    }
+  }
```

Comment:
Not sure that we need this, if we always get data from cache on `getTopPoolsForToken`

Reply:
We set tokens for each `tokensIntervalMs` in redis from init service. Don't really understand why we need to read tokens twice if we can read only in `getTopPoolsForToken`

Reply:
I think it's easier to get tokens from cache `getTopPoolsForToken` instead of using `updatePoolState` (might have less redis calls in general then), but current implementation should also work